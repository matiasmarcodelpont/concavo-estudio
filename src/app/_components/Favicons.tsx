'use client'

import { usePathname } from 'next/navigation'

const Favicons = () => {
  const pathname = usePathname()

  if (!pathname.includes('casa')) {
    return (
      <>
        <link rel='icon' href='/manifest/favicon_concavo.ico' sizes='any' />
        <link rel='icon' href='/manifest/icon_concavo?<generated>' type='image/<generated>' sizes='<generated>' />
        <link
          rel='apple-touch-icon'
          href='/manifest/apple-icon_concavo?<generated>'
          type='image/<generated>'
          sizes='<generated>'
        />
      </>
    )
  }

  return (
    <>
      <link rel='icon' href='/manifest/favicon_casa.ico' sizes='any' />
      <link rel='icon' href='/manifest/icon_casa?<generated>' type='image/<generated>' sizes='<generated>' />
      <link
        rel='apple-touch-icon'
        href='/manifest/apple-icon_casa?<generated>'
        type='image/<generated>'
        sizes='<generated>'
      />
    </>
  )
}

export default Favicons
