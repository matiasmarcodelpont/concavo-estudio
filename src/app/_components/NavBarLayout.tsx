'use client'

import { Drawer, DrawerContent } from '@/components/ui/drawer'
import { OmitReferences, Room } from '@/data/types'
import Link from 'next/link'
import { ReactNode, useCallback, useEffect, useState } from 'react'

export const NavBarLayout = ({
  rooms,
  children,
  opened = false,
}: {
  rooms: OmitReferences<Room>[]
  children?: ReactNode
  opened?: boolean
}) => {
  const [drawerOpen, setDrawerOpen] = useState(opened)
  useEffect(() => {
    setDrawerOpen(opened)
  }, [opened])

  const closeDrawer = useCallback(() => {
    setDrawerOpen(false)
  }, [])

  const links = [
    {
      text: 'C贸ncavo',
      href: '/',
      links: [{ text: 'Productos C贸ncavo', href: '/#productos_concavo' }],
    },
    {
      text: 'Casa C贸ncavo',
      href: '/casa',
      links: rooms.map(({ slug, name }) => ({ text: name, href: `/casa/${slug}` })),
    },
    {
      text: 'Nosotros',
      href: '/nosotros',
    },
  ]

  return (
    <Drawer direction='left' open={drawerOpen} onOpenChange={setDrawerOpen}>
      <DrawerContent className='pl-6 pr-7 uppercase overflow-y-auto'>
        <header className='border-black border-b mb-4'>
          <h1 className='my-4 font-bold text-xl'>C贸ncavo</h1>
        </header>

        <nav>
          <ul>
            {links.map(({ text, href, links }) => (
              <li key={href} className='my-4'>
                <Link href={href} onClick={closeDrawer} className='hover:underline'>
                  {text}
                </Link>

                {links && (
                  <ul className='ml-4'>
                    {links.map(({ text, href }) => (
                      <li key={href} className='my-2'>
                        <Link href={href} onClick={closeDrawer} className='hover:underline'>
                          {text}
                        </Link>
                      </li>
                    ))}
                  </ul>
                )}
              </li>
            ))}
          </ul>
        </nav>
      </DrawerContent>

      {children}
    </Drawer>
  )
}
