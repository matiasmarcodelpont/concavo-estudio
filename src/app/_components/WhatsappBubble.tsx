'use client'

import WhatsappIcon from '@/components/icons/WhatsappIcon'
import { cn } from '@/lib/utils'
import Link from 'next/link'
import React, { useState } from 'react'
import { Waypoint } from 'react-waypoint'

const MESSAGE = encodeURIComponent(
  '¡Hola! Estoy interesado en los productos que ofrecen. ¿Podrían darme más información?',
)

const WhatsappBubble = () => {
  const [show, setShow] = useState(false)

  return (
    <>
      <Link
        className={cn(
          'fixed right-3 bottom-2 sm:right-4 sm:bottom-3 md:right-5 md:bottom-4 lg:right-6 lg:bottom-5 xl:right-7 bg-bone p-2 rounded-full hover:scale-105 transition-all z-50 shadow-lg',
          show ? 'visible opacity-80' : 'invisible opacity-0',
        )}
        target='_blank'
        href={`https://wa.me/5491136111800?text=${MESSAGE}`}
      >
        <WhatsappIcon className='w-9 h-9 fill-dark-gray cursor-pointer' />
      </Link>
      <div className='absolute top-[150vh] right-0'>
        <Waypoint
          onEnter={() => {
            setShow(true)
          }}
        />
      </div>
    </>
  )
}

export default WhatsappBubble
