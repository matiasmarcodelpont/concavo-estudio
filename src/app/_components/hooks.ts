'use client'

import { usePathname } from 'next/navigation'
import { useCallback, useEffect, useRef, useState } from 'react'
import { QRCode } from 'react-qrcode-logo'

export const useHandleQr = () => {
  const [clickCount, setClickCount] = useState(0)
  const qrRef = useRef<QRCode>(null)

  const pathname = usePathname()
  const currentUrl = `${window.location.origin}${pathname}`

  const handleClick = () => {
    setClickCount((prev) => prev + 1)
  }

  const handleTripleClick = useCallback(() => {
    qrRef.current?.download('png', pathname === '/' ? 'home' : pathname.slice(1))
  }, [pathname])

  useEffect(() => {
    if (clickCount === 3) {
      handleTripleClick()
      setClickCount(0)
    }

    const timeout = setTimeout(() => {
      setClickCount(0)
    }, 500)

    return () => {
      clearTimeout(timeout)
    }
  }, [clickCount, handleTripleClick])

  return { currentUrl, handleClick, qrRef }
}
