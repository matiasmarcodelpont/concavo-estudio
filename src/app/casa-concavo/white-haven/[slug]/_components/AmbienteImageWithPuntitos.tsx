'use client'

import { ImageWithResolvedPuntitos, ResolvedPuntito } from '@/data/repositories/ambientes'
import { Ambiente, AmbienteImage } from '@/data/types'
import Image from 'next/image'
import Link from 'next/link'
import { useEffect, useMemo, useRef, useState } from 'react'

const PuntitoLink = ({
  puntito,
  imageSize,
  renderedImageSize,
}: {
  puntito: ResolvedPuntito
  imageSize: Pick<AmbienteImage, 'width' | 'height'>
  renderedImageSize: Pick<AmbienteImage, 'width' | 'height'>
}) => {
  const { top, left } = useMemo(() => {
    const { x, y } = puntito.coordinates
    const { width, height } = imageSize
    const { width: renderedWidth, height: renderedHeight } = renderedImageSize

    const left = (x * renderedWidth) / width
    const top = (y * renderedHeight) / height

    return {
      left: `${left.toString()}px`,
      top: `${top.toString()}px`,
    }
  }, [puntito.coordinates, imageSize, renderedImageSize])

  return (
    <Link
      href={`/productos/${puntito.producto.slug}`}
      className='absolute animate-fade-in delay-1000 -translate-x-1/2 -translate-y-1/2'
      style={{ top, left }}
    >
      <div className='size-8 rounded-full bg-bone absolute -top-4 -left-4 animate-puntito-outer' />
      <div className='size-4 rounded-full bg-bone absolute -translate-x-1/2 -translate-y-1/2' />
    </Link>
  )
}

export const AmbienteImageWithPuntitos = ({
  ambiente,
  image,
}: {
  ambiente: Omit<Ambiente, 'productos' | 'images'>
  image: ImageWithResolvedPuntitos
}) => {
  const imageRef = useRef<HTMLImageElement>(null)

  const [{ width, height }, setImageRenderedDimensions] = useState({ width: 0, height: 0 })

  useEffect(() => {
    if (!imageRef.current) {
      return
    }

    setImageRenderedDimensions({
      width: imageRef.current.clientWidth,
      height: imageRef.current.clientHeight,
    })
  }, [])

  return (
    <>
      <Image
        ref={imageRef}
        src={`/ambientes/${ambiente.slug}/${image.src}`}
        alt={ambiente.name}
        width={image.width}
        height={image.height}
        className='size-full object-cover'
      />

      {image.puntitos.map((puntito) => (
        <PuntitoLink
          key={puntito.producto.slug}
          puntito={puntito}
          imageSize={{ width: image.width, height: image.height }}
          renderedImageSize={{ width, height }}
        />
      ))}
    </>
  )
}
