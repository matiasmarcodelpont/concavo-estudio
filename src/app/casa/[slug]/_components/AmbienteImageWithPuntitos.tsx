'use client'

import { ImageWithResolvedPuntitos, ResolvedPuntito } from '@/data/repositories/ambientes'
import { Ambiente, AmbienteImage } from '@/data/types'
import Image from 'next/image'
import Link from 'next/link'
import { useEffect, useMemo, useRef, useState } from 'react'

const PuntitoLink = ({
  puntito,
  imageSize,
  renderedImageSize,
}: {
  puntito: ResolvedPuntito
  imageSize: Pick<AmbienteImage, 'width' | 'height'>
  renderedImageSize: Pick<AmbienteImage, 'width' | 'height'>
}) => {
  const { top, left } = useMemo(() => {
    const { x, y } = puntito.coordinates
    const { width, height } = imageSize
    const { width: renderedWidth, height: renderedHeight } = renderedImageSize

    const left = (x * renderedWidth) / width
    const top = (y * renderedHeight) / height

    return {
      left: `${left.toString()}px`,
      top: `${top.toString()}px`,
    }
  }, [puntito.coordinates, imageSize, renderedImageSize])

  return (
    <Link
      href={`/productos/${puntito.producto.slug}`}
      className='absolute size-8 p-2 rounded-full bg-black/20 hover:scale-125 transition -translate-x-1/2 -translate-y-1/2'
      style={{ top, left }}
    >
      <div className='size-full rounded-full bg-white' />
    </Link>
  )
}

export const AmbienteImageWithPuntitos = ({
  ambiente,
  image,
}: {
  ambiente: Omit<Ambiente, 'productos' | 'images'>
  image: ImageWithResolvedPuntitos
}) => {
  const imageRef = useRef<HTMLImageElement>(null)

  const [{ width, height }, setImageRenderedDimensions] = useState({ width: 0, height: 0 })

  useEffect(() => {
    if (!imageRef.current) {
      return
    }

    setImageRenderedDimensions({
      width: imageRef.current.clientWidth,
      height: imageRef.current.clientHeight,
    })
  }, [])

  return (
    <>
      <Image
        ref={imageRef}
        src={`/ambientes/${ambiente.slug}/${image.src}`}
        alt={ambiente.name}
        width={image.width}
        height={image.height}
        className='size-full object-cover'
      />

      {image.puntitos.map((puntito) => (
        <PuntitoLink
          key={puntito.producto.slug}
          puntito={puntito}
          imageSize={{ width: image.width, height: image.height }}
          renderedImageSize={{ width, height }}
        />
      ))}
    </>
  )
}
