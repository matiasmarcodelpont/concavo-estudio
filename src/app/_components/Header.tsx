'use client'

import Logo from '@/components/domain/Logo'
import Link from 'next/link'
import { NavBarLayout } from './NavBarLayout'
import { Ambiente, OmitReferences } from '@/data/types'
import { usePathname } from 'next/navigation'
import { cn } from '@/lib/utils'

interface HeaderProps {
  ambientes: OmitReferences<Omit<Ambiente, 'images'>>[]
}

export const Header = ({ ambientes }: HeaderProps) => {
  const pathname = usePathname()

  return (
    <header className='fixed top-0 w-full bg-bone flex justify-between px-2 sm:px-6 py-4 items-center z-10'>
      <NavBarLayout ambientes={ambientes} />

      <h1 className={cn(pathname == '/' ? 'invisible' : '', 'my-0')}>
        <Link href='/'>
          <Logo />
        </Link>
      </h1>

      <div className='w-[40px]' />
    </header>
  )
}
