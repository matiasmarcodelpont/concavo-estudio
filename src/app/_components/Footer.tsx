'use client'

import { Qr } from '@/components/ui/qr'
import { ambientesRepository, productosRepository } from '@/controllers'
import { useHandleQr } from './hooks'
import LogoAndSocialMedia from '@/components/domain/LogoAndSocialMedia'

export const Footer = () => {
  const ambientes = ambientesRepository.getAmbientes()
  const productos = productosRepository.getProductosConcavo()

  const { currentUrl, handleClick, qrRef } = useHandleQr()

  const siteMap = [
    {
      name: 'Home',
      links: [
        {
          name: 'Cóncavo',
          slug: 'concavo',
          href: '/',
        },
        {
          name: 'Productos Cóncavo',
          slug: 'productos-concavo',
          href: '/#productos-concavo',
        },
        {
          name: 'Casa Cóncavo',
          slug: 'casa-concavo',
          href: '/casa-concavo',
        },
      ],
    },
    {
      name: 'Casa Concavo',
      links: ambientes.map(({ name, slug }) => ({ name, slug, href: `/casa-concavo/${slug}` })),
    },
    {
      name: 'Productos',
      links: productos.map(({ name, slug }) => ({ name, slug, href: `/productos/${slug}` })),
    },
  ]

  return (
    <footer className='bg-white text-black flex gap-5 md:gap-10 px-6 sm:px-10 md:px-12 pt-10 pb-6 flex-wrap-reverse'>
      {siteMap.map(({ name, links }, index) => (
        <section key={`sitemap-${index.toString()}-${name}`} className='hidden md:block'>
          <h1 className='text-base lg:text-xl font-tt-norms font-bold uppercase mb-2'>{name}</h1>
          <ul className='text-xs gap-y-2 gap-x-6 flex flex-col flex-wrap max-h-[150px]'>
            {links.map(({ name, slug, href }) => (
              <li key={slug} className='hover:underline'>
                <a href={href}>{name}</a>
              </li>
            ))}
          </ul>
        </section>
      ))}

      <section className='flex gap-6 xs:gap-10 sm:gap-14 ml-auto mr-auto md:mr-0 flex-wrap'>
        <LogoAndSocialMedia variant='black' className='mx-auto' />
        <div className='mx-auto'>
          <Qr value={currentUrl} onClick={handleClick} />
        </div>
        <div className='absolute left-[-100vw] top-0'>
          <Qr ref={qrRef} value={currentUrl} size={300} />
        </div>
      </section>
    </footer>
  )
}
