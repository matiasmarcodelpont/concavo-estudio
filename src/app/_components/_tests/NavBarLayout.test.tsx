import { render } from '@testing-library/react'
import { NavBarLayout } from '@/app/_components/NavBarLayout'

describe('NavBarLayour', () => {
  it('renders links to all pages and to each ambiente', () => {
    const { getByRole } = render(
      <NavBarLayout
        opened
        ambientes={[
          { name: 'Ambiente 1', slug: 'ambiente-1' },
          { name: 'Ambiente 2', slug: 'ambiente-2' },
          { name: 'Ambiente 3', slug: 'ambiente-3' },
        ]}
      />,
    )

    expect(getByRole('navigation')).toMatchSnapshot()

    const homeLink = getByRole('link', { name: 'Cóncavo' }) as HTMLAnchorElement
    expect(homeLink.href).toMatch(/^https?:\/\/[^/]+\/$/)

    const concavoProductosLink = getByRole('link', { name: 'Productoos Cóncavo' }) as HTMLAnchorElement
    expect(concavoProductosLink.href).toMatch(/^https?:\/\/[^/]+\/#productoos_concavo$/)

    const concavoHomeLink = getByRole('link', { name: 'Casa Cóncavo' }) as HTMLAnchorElement
    expect(concavoHomeLink.href).toMatch(/^https?:\/\/[^/]+\/casa$/)

    const ambiente1Link = getByRole('link', { name: 'Ambiente 1' }) as HTMLAnchorElement
    expect(ambiente1Link.href).toMatch(/^https?:\/\/[^/]+\/casa\/ambiente-1$/)

    const ambiente2Link = getByRole('link', { name: 'Ambiente 2' }) as HTMLAnchorElement
    expect(ambiente2Link.href).toMatch(/^https?:\/\/[^/]+\/casa\/ambiente-2$/)

    const ambiente3Link = getByRole('link', { name: 'Ambiente 3' }) as HTMLAnchorElement
    expect(ambiente3Link.href).toMatch(/^https?:\/\/[^/]+\/casa\/ambiente-3$/)

    const aboutLink = getByRole('link', { name: 'Nosotros' }) as HTMLAnchorElement
    expect(aboutLink.href).toMatch(/^https?:\/\/[^/]+\/nosotros$/)
  })
})
