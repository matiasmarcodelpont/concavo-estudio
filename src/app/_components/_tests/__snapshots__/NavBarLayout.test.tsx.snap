// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBarLayour renders links to all pages and to each room 1`] = `
<nav>
  <ul>
    <li
      class="my-4"
    >
      <a
        class="hover:underline"
        href="/"
      >
        Cóncavo
      </a>
      <ul
        class="ml-4"
      >
        <li
          class="my-2"
        >
          <a
            class="hover:underline"
            href="/#productos_concavo"
          >
            Productos Cóncavo
          </a>
        </li>
      </ul>
    </li>
    <li
      class="my-4"
    >
      <a
        class="hover:underline"
        href="/casa"
      >
        Casa Cóncavo
      </a>
      <ul
        class="ml-4"
      >
        <li
          class="my-2"
        >
          <a
            class="hover:underline"
            href="/casa/room-1"
          >
            Room 1
          </a>
        </li>
        <li
          class="my-2"
        >
          <a
            class="hover:underline"
            href="/casa/room-2"
          >
            Room 2
          </a>
        </li>
        <li
          class="my-2"
        >
          <a
            class="hover:underline"
            href="/casa/room-3"
          >
            Room 3
          </a>
        </li>
      </ul>
    </li>
    <li
      class="my-4"
    >
      <a
        class="hover:underline"
        href="/nosotros"
      >
        Nosotros
      </a>
    </li>
  </ul>
</nav>
`;
